src/apps/EDGE3D-Cuda.cpp

```diff
@@ -0,0 +1,112 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-23, Lawrence Livermore National Security, LLC
+// and RAJA Performance Suite project contributors.
+// See the RAJAPerf/LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "EDGE3D.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+#if defined(RAJA_ENABLE_CUDA)
+
+#include "common/CudaDataUtils.hpp"
+
+#include "AppsData.hpp"
+
+#include <iostream>
+
+namespace rajaperf
+{
+namespace apps
+{
+
+template < size_t block_size >
+__launch_bounds__(block_size)
+__global__ void edge3d(Real_ptr sum,
+                       const Real_ptr x0, const Real_ptr x1,
+                       const Real_ptr x2, const Real_ptr x3,
+                       const Real_ptr x4, const Real_ptr x5,
+                       const Real_ptr x6, const Real_ptr x7,
+                       const Real_ptr y0, const Real_ptr y1,
+                       const Real_ptr y2, const Real_ptr y3,
+                       const Real_ptr y4, const Real_ptr y5,
+                       const Real_ptr y6, const Real_ptr y7,
+                       const Real_ptr z0, const Real_ptr z1,
+                       const Real_ptr z2, const Real_ptr z3,
+                       const Real_ptr z4, const Real_ptr z5,
+                       const Real_ptr z6, const Real_ptr z7,
+                       Index_type ibegin, Index_type iend)
+{
+  Index_type i = ibegin + blockIdx.x * block_size + threadIdx.x;
+  if (i < iend) {
+    EDGE3D_BODY;
+  }
+}
+
+
+template < size_t block_size >
+void EDGE3D::runCudaVariantImpl(VariantID vid)
+{
+  const Index_type run_reps = getRunReps();
+  const Index_type ibegin = m_domain->fpz;
+  const Index_type iend = m_domain->lpz+1;
+
+  EDGE3D_DATA_SETUP;
+
+  if ( vid == Base_CUDA ) {
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      const size_t grid_size = RAJA_DIVIDE_CEILING_INT(iend, block_size);
+
+      edge3d<block_size><<<grid_size, block_size>>>(sum,
```

Comment:
Need to add stream/resource use to all of the cuda/hip kernels, see the other cuda/hip kernels for an example.

Reply:
Done